### PR TITLE
Add Dynamic Breadth First Scan Ordering to GC

### PR DESCRIPTION
**Pull request by jonoommen**
Add a gc scavenger scan ordering feature that enables the
copying of a hot field marked by the JIT immediately after the
object containing the hot field is copied.

Issue: eclipse/openj9#7552
Signed-off-by: Jonathan Oommen jon.oommen@gmail.com

### CODE REVIEW

**Comment by amicic** [comment]
A follow up item: in concurrent copying variant we may want to disable the optimization during root scanning, as to minimize the scan (copy) time just to directly referenced roots.

File: gc/base/standard/Scavenger.cpp
Comment (amicic) [inline gc/base/standard/Scavenger.cpp:1656]: I believe we can avoid this check and rely on _hotFieldCopyDepthCount being fixed to its max, in case of other ordering schemes.
If so, it would be wise to pull out the depth check outside of depthCopyHotFields in case the compiler does not obey inlining

**Comment by amicic** [comment]
This PR is using (C language) recursion to implement a limited (up to 4 levels by default) nested step of the optimizations. While we are generally avoid hard to rely on recursion (this might very well be the first one in GC code0, but being that limited and the methods using just a few scalar variables on stack, I'm ok to rely on it, for the sake of code simplicity (otherwise, we'd have to implement/maintain an explicit stack, which seems like overkill).

File: gc/base/standard/Scavenger.cpp
Comment (jonoommen) [reply gc/base/standard/Scavenger.cpp:1656]: In order to rely on _hotFieldCopyDepthCount, I would need to wrap my calls to enable and disable depth copying during the finalizable pass and phantom references pass with the mentioned check for dynamic breadth first so that the depth does not get modified during the finalizable and phantom references pass, and then would just need to initially set the depth to max by default and then if dynamic breadth first scan ordering is enabled via command line, set the _hotFieldCopyDepthCount to zero. Would this solution be better than having this check here? @amicic

Comment (jonoommen) [reply gc/base/standard/Scavenger.cpp:1656]: @amicic

Comment (jonoommen) [reply gc/base/standard/Scavenger.cpp:1656]: Resolved via webex discussion.

Comment (amicic) [reply gc/base/standard/Scavenger.cpp:1656]: I see now the check added to enableHotFieldDepthCopy(), but I expected the check would be removed from here!?

Comment (jonoommen) [reply gc/base/standard/Scavenger.cpp]: Yes my mistake, I ensured the check was not needed but forgot to remove it before pushing the changes.

**Comment by amicic** [comment]
@andrewcraik I have reviewed and  approved GC part of changes. Have you (or anybody else) had any comments for Compiler part?

File: compiler/optimizer/VPHandlers.cpp
Comment (andrewcraik) [inline compiler/optimizer/VPHandlers.cpp:2780]: this should be deleted and not commented out

**Comment by andrewcraik** [comment]
Beyond the commented out code I think this is reasonable. It feels a bit strange having all the options and threshold in OMR and the opt pass to do the marking in OpenJ9, but some of this can't be avoided at present due to the limitations in the compiler option parsing infrastructure. I think this is fine for now.

**Review by andrewcraik — APPROVED** [review]

**Comment by andrewcraik** [comment]
@charliegracie @youngar @rwy0717 - I think a code owner for the GC needs to be ok before I merge. @amicic has reviewed and approved.

File: gc/base/standard/Scavenger.hpp
Comment (youngar) [inline gc/base/standard/Scavenger.hpp:271]:
```suggestion
	 * @param forwardedHeader Forwarded header of an object
	 * @param destinationObjectPtr DestinationObjectPtr of the object described by the forwardedHeader
```

File: gc/base/standard/Scavenger.hpp
Comment (youngar) [inline gc/base/standard/Scavenger.hpp:278]:
```suggestion
	 * @param destinationObjectPtr The object who's hot field will be copied
	 * @param offset The object field offset of the hot field to be copied 
```

File: gc/base/standard/Scavenger.cpp
Comment (youngar) [inline gc/base/standard/Scavenger.cpp:1704]:
```suggestion
	omrobjectptr_t objectPtr = hotFieldObject.readReferenceFromSlot();
```

File: example/glue/ObjectModelDelegate.hpp
Comment (youngar) [inline example/glue/ObjectModelDelegate.hpp:215]: I think in general OMR tries to use standard integer types instead of `U_8`.
```suggestion
	MMINLINE uint8_t
	getHotFieldOffset(MM_ForwardedHeader *forwardedHeader)
	{
		return UINT8_MAX;
	}
```

@amicic @jonoommen Should this glue API return a `uintptr_t` for field offsets?

**Review by youngar — CHANGES_REQUESTED** [review]
A couple nitpicks

File: example/glue/ObjectModelDelegate.hpp
Comment (jonoommen) [reply example/glue/ObjectModelDelegate.hpp:215]: For field offsets I have used 'uint8' everywhere. I don't have a preference as both seem correct to me. Any thoughts @amicic?

Comment (amicic) [reply example/glue/ObjectModelDelegate.hpp:215]: As far as the width is concerned we should be using the widest offset possible out of all downstream projects. OpenJ9 is the only one implementing this, and from what I can see the info comes from getInstanceFieldOffset() that returns uint32_t.

As far as using U_32 vs uint32_t (or any other width) we should be using the latter (I think it's the general trend in OMR code).

Comment (jonoommen) [reply example/glue/ObjectModelDelegate.hpp:215]: Looking at this now after our offline discussion, shall I leave this as is? @youngar

Comment (amicic) [reply example/glue/ObjectModelDelegate.hpp]: I'm fine with limiting the offset within uint8_t, since it's unknown if there would be any benefit with large field offsets.

As far as downstream projects are concerned, OpenJ9 while may have field offset larger than uint_8, for this particular purpose we made sure anything above that threshold will be ignored - so there is no correctness or safety issue.

**Comment by youngar** [comment]
@genie-omr build all

**Review by youngar — APPROVED** [review]

**Pull request merged**
